onnxruntime/core/optimizer/shape_input_merge.cc

```diff
@@ -0,0 +1,76 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+#include "core/optimizer/shape_input_merge.h"
+
+#include "core/graph/graph_utils.h"
+
+namespace onnxruntime {
+
+namespace {
+std::string GetShapeString(const NodeArg* input_arg) {
+  auto shape = input_arg->Shape();
+  if (!shape) return "";
+  std::string shape_str = "[";
+  for (int i = 0; i < shape->dim_size(); ++i) {
+    if (i != 0) shape_str += ",";
+    auto dim = shape->dim(i);
+    if (dim.has_dim_value()) {
+      shape_str += std::to_string(dim.dim_value());
+    } else if (dim.has_dim_param()) {
+      shape_str += ("'" + dim.dim_param() + "'");
+    } else {
+      return "";
+    }
+  }
+  shape_str += "]";
+  return shape_str;
+}
+}  // namespace
+
+Status ShapeInputMerge::ApplyImpl(Graph& graph, bool& modified, int graph_level, const logging::Logger& logger) const {
+  GraphViewer graph_viewer(graph);
+  const auto& node_topology_list = graph_viewer.GetNodesInTopologicalOrder();
+  InlinedHashMap<std::string, InlinedVector<Node*>> input_hash_to_nodes;
+  for (auto node_index : node_topology_list) {
+    auto* p_node = graph.GetNode(node_index);
+    if (!p_node) continue;  // we removed the node as part of an earlier fusion
+    ORT_RETURN_IF_ERROR(Recurse(*p_node, modified, graph_level, logger));
```

Comment:
what's the purpose of Recurse here?

Reply:
I guess this can also be applied to sub-graph. Most of the optimizers call this, including the CSE fusion.